dttools/src/bucketing_fast.c

```diff
@@ -0,0 +1,240 @@
+#include "bucketing_fast.h"
+#include "bucketing.h"
+#include "random.h"
+#include "list.h"
+
+double bucketing_fast_predict(double prev_val, bucketing_state* s)
+{
+    if (s->in_sampling_phase)
+    {
+        return s->default_value;
+    }
+    double rand = random_double();
+    
+    struct list_cursor* lc = list_cursor_create(s->sorted_buckets);
+    list_seek(lc, 0);
+        
+    bucketing_bucket** bb_ptr = malloc(sizeof(*bb_ptr));
+    
+    double sum = 0;
+    double ret_val;
+
+    for (unsigned int i = 0; i < list_length(s->sorted_buckets); ++i, list_next(lc))
+    {    
+        list_get(lc, (void**) bb_ptr);
+        
+        if (i == list_length(s->sorted_buckets) - 1)
+        {
+            ret_val = (*bb_ptr)->val;
+            list_cursor_destroy(lc);
+            free(bb_ptr);
+            return ret_val;
+        }
+        
+        sum += (*bb_ptr)->prob;
+        
+        if (sum > rand)
+        {
+            ret_val = (*bb_ptr)->val;
+            list_cursor_destroy(lc);
+            free(bb_ptr);
+            return ret_val;
+        }
+    }
+
+    return -1;  //control should never reach here
+}
+
+int bucketing_fast_update_buckets(bucketing_state* s)
+{
+    list_free(s->sorted_buckets);
+    list_delete(s->sorted_buckets);
+
+    struct list* break_point_list = bucketing_find_break_points(s);
+    printf("Done finding break points\n");
+    bucketing_cursor_w_pos* tmp_break_point;
+    bucketing_bucket* tmp_bucket;
+
+    bucketing_point** tmp_point_ptr = malloc(sizeof(*tmp_point_ptr));
+    int prev_pos = 0;
+    s->sorted_buckets = list_create();
+    list_first_item(break_point_list);
+
+    while ((tmp_break_point = list_next_item(break_point_list)) != NULL)
+    {
+        list_get(tmp_break_point->lc, (void**) tmp_point_ptr);
+        tmp_bucket = bucketing_bucket_create((*tmp_point_ptr)->val, (tmp_break_point->pos - prev_pos + 1)/list_length(s->sorted_points));
+        prev_pos = tmp_break_point->pos;
+        list_push_tail(s->sorted_buckets, tmp_bucket);
+    }
+    
+    printf("Done pushing new buckets\n");
+    printf("Break point list length %u\n", list_length(break_point_list));
+    list_clear(break_point_list, (void*) bucketing_cursor_w_pos_delete);
+    printf("Done dropping all points in break point list\n");
+    list_delete(break_point_list);
+    free(tmp_point_ptr);
+    return 0;
+}
+
+struct list* bucketing_find_break_points(bucketing_state* s)
+{
+    int min = 0;
+    int max = list_length(s->sorted_points) - 1;
+    struct list* break_point_list = list_create();
+
+    /* create list and push (0, n-1) to list */
+    struct list* bucket_range_list = list_create();
+    bucketing_bucket_range* init_range = bucketing_bucket_range_create(min, max, s->sorted_points);
+    list_push_tail(bucket_range_list, init_range);
+
+    bucketing_bucket_range* lo_bucket_range;
+    bucketing_bucket_range* hi_bucket_range;
+    bucketing_cursor_w_pos** break_point = malloc(sizeof(*break_point));
```

Comment:
I think this code will be easier to read/write if you remove a level of indirection.  Try this:
`
bucketing_cursor_w_pos * break_point = 0;
list_get(lc,&break_point);
 . . .break_point->pos . . .
```